warp/yul/DeadcodeEliminator.py

```diff
@@ -0,0 +1,172 @@
+    - ForLoopIliminator
+    """
+
+    def map(self, node: ast.Node, **kwargs) -> ast.Node:
+        (vs, node) = self.visit(node)
+        assert not vs, "There should be no uninstantiated variables in the top scope."
+        return node
+
+    def visit_list(
+        self, nodes: Sequence[ast.Node], *args, **kwargs
+    ) -> (LiveVars, Tuple[ast.Node]):
+        if len(nodes) == 0:
+            return (set(), list())
+        lis = [self.visit(x, *args, **kwargs) for x in nodes]
+        [var_set, nodes] = list(map(list, zip(*lis)))
+        live_vars = reduce(lambda s1, s2: s1 | s2, var_set, fset())
+        return (live_vars, nodes)
+
+    def visit_literal(self, node: ast.Literal) -> Tuple[LiveVars, ast.Literal]:
+        return (fset(), node)
+
+    def visit_identifier(self, node: ast.Identifier) -> Tuple[LiveVars, ast.Identifier]:
+        return (fset({node.name}), node)
+
+    def visit_typed_name(self, node: ast.TypedName):
+        return (fset({node.name}), node)
+
+    def visit_function_call(
+        self, node: ast.FunctionCall
+    ) -> (LiveVars, ast.FunctionCall):
+        (live_vars, arguments) = self.visit(node.arguments)
+        return (
+            live_vars,
+            ast.FunctionCall(function_name=node.function_name, arguments=arguments),
+        )
+
+    def visit_expression_statement(
+        self, node: ast.ExpressionStatement
+    ) -> (LiveVars, ast.ExpressionStatement):
+        (live_vars, _) = self.visit(node.expression)
+        return (live_vars, node)
+
+    def visit_assignment(self, node: ast.Assignment) -> Tuple[LiveVars, ast.Assignment]:
+        (live_vars, _) = self.visit(node.value)
+        return (live_vars, node)
+
+    def visit_variable_declaration(
+        self, node: ast.VariableDeclaration
+    ) -> (LiveVars, ast.VariableDeclaration):
+        if node.value is None:
+            return (fset(), node)
+        (live_vars, _) = self.visit(node.value)
+        return (live_vars, node)
+
+    def visit_function_definition(
+        self, node: ast.FunctionDefinition
+    ) -> (LiveVars, ast.FunctionDefinition):
+        # We need to ignore the generated getters from kudu since their
+        # body is only instantiated in ToCairoVisitor
+        if node.name.startswith("getter"):
+            return (fset(), node)
+        (live_vars, body) = self.visit(
+            node.body, return_vars=fset({n.name for n in node.return_variables})
+        )
+        live_vars -= fset({arg.name for arg in node.parameters})
+        return (
+            live_vars,
+            ast.FunctionDefinition(
+                name=node.name,
+                parameters=node.parameters,
+                return_variables=node.return_variables,
+                body=body,
+            ),
+        )
+
+    def visit_if(
+        self, node: ast.If, live_vars: Set[str] = None, return_vars: Set[str] = None
+    ) -> Tuple[LiveVars, ast.If]:
+        if any(
+            isinstance(statement, ast.Leave)
```

Comment:
There can be nested blocks and nested ifs, so it's probably not enough of a check. I suggest storing `return_vars` as a class field in a way similar to the `ScopeFlattener._new_function` context manager. Would be nice to wrap it in a ReaderT, but alas we're in the python land.

Reply:
Great idea thank you